### PR TITLE
New version of multipart_form

### PR DESCRIPTION
**Pull request by dinosaure**
This PR does not change the final API but the internal layout of the library. It took same update from `mrmime` and use `unstrctrd` as a parser for headers. I deleted some dependence such as `rosetta`. The next step is to add some tests and may be think about a release.

### CODE REVIEW

**Comment by dinosaure** [comment]
Good ratio!

File: lib/multipart_form.ml [hunk 1]
```diff
@@ -1,18 +1,9 @@
...
-and 'id atom =
-  { fields : Field.field list
-  ; contents : 'id t option }
+open Stdlib
 
-type 'id emitters = Field.field list -> (string option -> unit) * 'id
+module Field = Field
+module Header = Header
+module Content_type = Content_type
+module Content_disposition = Content_disposition
```
Comment (anmonteiro) [inline lib/multipart_form.ml:6]: Can we also export `Content_encoding`?

File: lib/multipart_form.ml [hunk 2]
```diff
@@ -199,141 +189,156 @@ module QP = struct
...
-
-  let to_end_of_input_with_push ?(end_of_line = "\n") push =
-    let write_data x = push (Some x) in
-    let write_line x = push (Some (x ^end_of_line)) in
-    to_end_of_input ~write_data ~write_line
 end
 
+type 'a elt = { header : Header.t; body : 'a }
+
+type 'a t = Leaf of 'a elt | Multipart of 'a t option list elt
```
Comment (anmonteiro) [inline lib/multipart_form.ml:267]: this signature looks a little different than before -- was that intentional? Feels like there's more nesting

**Review by anmonteiro — COMMENTED** [review]
Hey @dinosaure this refactoring looks great! Thanks for spending some time polishing this up. I left a suggestion and a minor question.

File: lib/multipart_form.ml (hunk 2 above)
Comment (dinosaure) [reply lib/multipart_form.ml:267]: Yes, I think the type is less complex and of course, it did not highlight real cases, but:
1) a part can contain an other multipart (`'a t`)
2) a part can be empty (`option`)
3) a multipart is a list of parts (`list`)
4) and a multipart contains a header (`elt`)

When I implemented tests, I discovered n useful function such as `'a t -> (string * 'a) list` (where the `string` is the name of the field - given by `Content-Disposition`) and facilitate the processing of a `multipart-form` for the end-user. I will try to improve the API next week.

**Comment by dinosaure** [comment]
So I think it's good to merge for the moment, @anmonteiro if you have any feedback about the API dans what I should provide to help you (and others) to parse the multipart-form, tell me. On my side, I think it's enough for my usage.

**Comment by anmonteiro** [comment]
Thank you!

**Comment by anmonteiro** [comment]
One more piece of feedback: it'd be cool to expose the pretty printers e.g. `Content_type.Type.pp`

**Pull request merged**